mailpoet/lib/NewsletterTemplates/TemplateImageLoader.php

```diff
@@ -0,0 +1,83 @@
+<?php // phpcs:ignore SlevomatCodingStandard.TypeHints.DeclareStrictTypes.DeclareStrictTypesMissing
+
+namespace MailPoet\NewsletterTemplates;
+
+use MailPoet\WP\Functions as WPFunctions;
+
+class TemplateImageLoader {
+  const TIMEOUT = 30; // seconds
+
+  /** @var WPFunctions */
+  private $wp;
+
+  public function __construct(
+    WPFunctions $wp
+  ) {
+    $this->wp = $wp;
+  }
+
+  /**
+   * Acts as a proxy for html2canvas
+   */
+  public function loadExternalImage(string $url) {
+    if (!$this->isUrlAllowed($url)) {
+      // URL not allowed
+      return false;
+    }
+    $image = $this->downloadUrl($url);
+    if ($this->wp->isWpError($image)) {
+      // Failed to load the image
+      return false;
+    }
+    if (!$this->isTypeAllowed($image, $mime)) {
+      // Wrong file type
+      @unlink($image);
+      return false;
+    }
+    header('Content-Type: ' . $mime);
+    readfile($image);
+    @unlink($image);
+    return true;
+  }
+
+  protected function downloadUrl($url) {
+    require_once ABSPATH . '/wp-admin/includes/file.php';
+    return download_url($url, self::TIMEOUT);
+  }
+
+  private function isUrlAllowed($url) {
+    $urlParts = parse_url($url);
+    $allowedExtensions = ['gif', 'png', 'jpg', 'jpeg'];
+    if (
+      !isset($urlParts['path'])
+      || !preg_match('/(' . join('|', $allowedExtensions) . ')$/', $urlParts['path'])
```

Comment:
To make this a bit more secure, I'd also check for the `.` before the extension. Not sure if it's exploitable, but otherwise it would allow `.insecure-jpg`.
```suggestion
      || !preg_match('/\.(' . join('|', $allowedExtensions) . ')$/', $urlParts['path'])
```

And on the other hand, make it a bit more permissive by ignoring case, `.JPG`, .`PNG`...
```suggestion
      || !preg_match('/(' . join('|', $allowedExtensions) . ')$/i', $urlParts['path'])
```

Reply:
Thanks! Done.